### DISCUSSION TITLE
Route not optimising correctly as Near locations are not optimising

### DISCUSSION THREAD

**Discussion by sharmas1ddharth** [TimefoldAI/timefold-solver#1278] [General]
I have 3 vehicles and 12 locations total and there were 4 locations near each other, all vehicles capacity is 5 but vehicle leaves one location only pick 3 locations that are near each other and goes to one far location then came back.

Attaching screenshot, can someone explain why is this happening and how to fix this as you can see blue vehicle pick three locations but leaves one location then red vehicle pick that location

![image](https://github.com/user-attachments/assets/2ca8989e-0c72-4405-95dc-69df836e419e)

**Comment by triceo** [comment]
The solver does not (and can not) guarantee that an optimal solution will be found. It gives you the best solution it can find, in the time it is given. The fact that a particular solution was not found is not a bug - it is a consequence of the stochastic nature of the algorithm used.

There are ways of helping the solver to find better solutions:
- Give it more run time.
- Increase move evaluation speed by writing faster constraints.
- Introduce additional custom moves that give the solver better options to sort through, ones that better understand the particular nature of your problem.

Since you are mentioning Nearby, which is a commercial feature of Timefold Solver Enterprise Edition, I assume you have access to Enterprise Edition - in that case, multi-threaded solving and automatic node-sharing will help you with speed.

**Reply by sharmas1ddharth** [reply]
But why it is not picking 4 locations in only one vehicle all the locations are very close to each other.

**Reply by triceo** [reply]
Because the solver did not find or select a move to make that situation happen. Maybe some constraint caused that situation to have a bad score. Or maybe the move to do this was filtered out. Or maybe the solver did not run for long enough to find that move.

There is no guarantee that the solver will find any particular solution.

**Reply by sharmas1ddharth** [reply]
Is there anything that I can do to find the cause, I am clueless actually

**Reply by greyhairredbear** [reply]
Since I'm being a month late this might not be an issue anymore, but it helped me in the past to manually create the desired solution (i.e. a solution with the 4 nearby locations on the same vehicle) and compare the score with the score of the solution returned by Timefold. That basically covers the second point (constraint causing bad score) @triceo pointed out and might give you some insights on other potential issues with constraints.
